### ISSUE TITLE
WebSocket TLS error when trying to connect IDE-R or SIDE-R

### ISSUE THREAD

**Issue by neg2led** [Ylianst/MeshCentral#2763]
Hi,

I have a fairly recently set up MeshCentral server and I'm not having any luck getting IDE-R connections to work :( Setup information:

- Ubuntu 20.04 LTS
- Kernel version: `5.8.0-1033-azure`
- Node.js LTS v14.17.0
- Not using MongoDB at the moment
- Agent group configured in 'Simple Admin Control Mode'
- ACM set up via USB key setup.bin

Agent and CIRA connections work fine, remote desktop works fine in both agent and hardware modes, but whenever I connect IDE-R or SIDE-R, MeshCentral itself is throwing some apparently-TLS-related websocket errors:

```
Jun 11 16:31:19 meshcentral node[27198]: Starting meshcentral syslog.
Jun 11 16:31:19 meshcentral meshcentral[27213]: MeshCentral v0.8.60 Server Start
Jun 11 16:31:19 meshcentral node[27198]: MeshCentral HTTP redirection server running on port 1080.
Jun 11 16:31:19 meshcentral node[27198]: MeshCentral v0.8.60, WAN mode.
Jun 11 16:31:20 meshcentral meshcentral[27213]: Server started
Jun 11 16:31:20 meshcentral node[27198]: MeshCentral Intel(R) AMT server running on server.domain.name:4433.
Jun 11 16:31:20 meshcentral node[27198]: MeshCentral HTTP server running on port 1443, alias port 443.
Jun 11 16:31:20 meshcentral node[27198]: Loaded web certificate from "https://server.domain.name:443/", host: "server.domain.name"
Jun 11 16:31:20 meshcentral node[27198]:   SHA384 cert hash: <snip>
Jun 11 16:31:20 meshcentral node[27198]:   SHA384 key hash: <snip>
Jun 11 16:31:27 meshcentral meshcentral[27213]: Account login
Jun 11 16:34:26 meshcentral node[27198]: AMT redir for itadmin to node//<node id>
Jun 11 16:34:26 meshcentral node[27198]: Opening Intel AMT CIRA transport connection to node//<node id>
Jun 11 16:34:27 meshcentral node[27198]: Intel AMT CIRA relay state change 2
Jun 11 16:34:27 meshcentral node[27198]: Intel AMT CIRA data 13
Jun 11 16:34:27 meshcentral node[27198]: Intel AMT CIRA data 10
Jun 11 16:34:27 meshcentral node[27198]: ReferenceError: i is not defined
Jun 11 16:34:27 meshcentral node[27198]:     at Object.obj.xxOnSocketData (/opt/meshcentral/node_modules/meshcentral/amt/amt-redir-mesh.js:330:28)
Jun 11 16:34:27 meshcentral node[27198]:     at Object.obj.forwardclient.onData (/opt/meshcentral/node_modules/meshcentral/amt/amt-redir-mesh.js:206:48)
Jun 11 16:34:27 meshcentral node[27198]:     at ProcessCommand (/opt/meshcentral/node_modules/meshcentral/mpsserver.js:984:59)
Jun 11 16:34:27 meshcentral node[27198]:     at TLSSocket.<anonymous> (/opt/meshcentral/node_modules/meshcentral/mpsserver.js:538:26)
Jun 11 16:34:27 meshcentral node[27198]:     at TLSSocket.emit (events.js:376:20)
Jun 11 16:34:27 meshcentral node[27198]:     at addChunk (internal/streams/readable.js:309:12)
Jun 11 16:34:27 meshcentral node[27198]:     at readableAddChunk (internal/streams/readable.js:280:11)
Jun 11 16:34:27 meshcentral node[27198]:     at TLSSocket.Readable.push (internal/streams/readable.js:223:10)
Jun 11 16:34:27 meshcentral node[27198]:     at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
Jun 11 16:34:57 meshcentral node[27198]: Intel AMT CIRA data 10
Jun 11 16:34:57 meshcentral node[27198]: ReferenceError: i is not defined
Jun 11 16:34:57 meshcentral node[27198]:     at Object.obj.xxOnSocketData (/opt/meshcentral/node_modules/meshcentral/amt/amt-redir-mesh.js:330:28)
Jun 11 16:34:57 meshcentral node[27198]:     at Object.obj.forwardclient.onData (/opt/meshcentral/node_modules/meshcentral/amt/amt-redir-mesh.js:206:48)
Jun 11 16:34:57 meshcentral node[27198]:     at ProcessCommand (/opt/meshcentral/node_modules/meshcentral/mpsserver.js:984:59)
Jun 11 16:34:57 meshcentral node[27198]:     at TLSSocket.<anonymous> (/opt/meshcentral/node_modules/meshcentral/mpsserver.js:538:26)
Jun 11 16:34:57 meshcentral node[27198]:     at TLSSocket.emit (events.js:376:20)
Jun 11 16:34:57 meshcentral node[27198]:     at addChunk (internal/streams/readable.js:309:12)
Jun 11 16:34:57 meshcentral node[27198]:     at readableAddChunk (internal/streams/readable.js:280:11)
Jun 11 16:34:57 meshcentral node[27198]:     at TLSSocket.Readable.push (internal/streams/readable.js:223:10)
Jun 11 16:34:57 meshcentral node[27198]:     at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
Jun 11 16:35:27 meshcentral node[27198]: Intel AMT CIRA data 10
Jun 11 16:35:27 meshcentral node[27198]: ReferenceError: i is not defined
Jun 11 16:35:27 meshcentral node[27198]:     at Object.obj.xxOnSocketData (/opt/meshcentral/node_modules/meshcentral/amt/amt-redir-mesh.js:330:28)
Jun 11 16:35:27 meshcentral node[27198]:     at Object.obj.forwardclient.onData (/opt/meshcentral/node_modules/meshcentral/amt/amt-redir-mesh.js:206:48)
Jun 11 16:35:27 meshcentral node[27198]:     at ProcessCommand (/opt/meshcentral/node_modules/meshcentral/mpsserver.js:984:59)
Jun 11 16:35:27 meshcentral node[27198]:     at TLSSocket.<anonymous> (/opt/meshcentral/node_modules/meshcentral/mpsserver.js:538:26)
Jun 11 16:35:27 meshcentral node[27198]:     at TLSSocket.emit (events.js:376:20)
Jun 11 16:35:27 meshcentral node[27198]:     at addChunk (internal/streams/readable.js:309:12)
Jun 11 16:35:27 meshcentral node[27198]:     at readableAddChunk (internal/streams/readable.js:280:11)
Jun 11 16:35:27 meshcentral node[27198]:     at TLSSocket.Readable.push (internal/streams/readable.js:223:10)
Jun 11 16:35:27 meshcentral node[27198]:     at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
Jun 11 16:35:32 meshcentral node[27198]: Intel AMT CIRA relay state change 0
```

Here's a sanitized copy of my `config.json`:

```
{
  "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json",
  "settings": {
    "cert": "server.domain.name",
    "_mongoDb": "mongodb://127.0.0.1:27017",
    "_mongoDbName": "meshcentral",
    "_mongoDbChangeStream": true,
    "WANonly": true,
    "_LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
    "_sessionSameSite": "strict",
    "_dbEncryptKey": "MyReallySecretPassword2",
    "_dbRecordsEncryptKey": "MyReallySecretPassword",
    "_dbRecordsDecryptKey": "MyReallySecretPassword",
    "__dbExpire": "Amount of time to keep various events in the database, in seconds. Below are the default values.",
    "_dbExpire": {
      "events": 1728000,
      "powerevents": 864000,
      "statsevents": 2592000
    },
    "port": 1443,
    "portBind": "127.0.0.1",
    "aliasPort": 443,
    "redirPort": 1080,
    "redirPortBind": "127.0.0.1",
    "redirAliasPort": 80,
    "_agentPort": 1234,
    "_agentPortBind": "127.0.0.1",
    "_agentAliasPort": 1234,
    "_agentAliasDNS": "agents.myserver.mydomain.com",
    "_agentPortTls": true,
    "_exactPorts": true,
    "allowLoginToken": false,
    "allowFraming": false,
    "_cookieIpCheck": false,
    "_cookieEncoding": "hex",
    "compression": true,
    "wscompression": false,
    "agentwscompression": false,
    "agentsInRam": false,
    "_webRTC": true,
    "nice404": true,
    "selfUpdate": false,
    "browserPing": 60,
    "browserPong": 60,
    "agentPing": 60,
    "agentPong": 120,
    "_agentIdleTimeout": 150,
    "meshErrorLogPath": "/opt/meshcentral/logs/error.log",
    "allowHighQualityDesktop": true,
    "desktopMultiplex": false,
    "authLog": "/opt/meshcentral/logs/auth.log",
    "manageAllDeviceGroups": [
      "user//itadmin"
    ],
    "manageCrossDomain": [
      "user//itadmin"
    ],
    "tlsOffload": "127.0.0.1,::1",
    "_trustedProxy": "127.0.0.1,::1",
    "mpsPort": 4433,
    "_mpsPortBind": "127.0.0.1",
    "_mpsAliasPort": 4433,
    "_mpsAliasHost": "server.domain.name",
    "_mpsTlsOffload": true,
    "_no2FactorAuth": true,
    "log": "main,web,webrequest,cert",
    "syslog": "meshcentral",
    "webrtcConfig": {
      "iceServers": [
        {
          "urls": "stun:stun.services.mozilla.com"
        },
        {
          "urls": "stun:stun.l.google.com:19302"
        }
      ]
    },
    "__maxInvalidLogin": "Time in minutes, max amount of bad logins from a source IP in the time before logins are rejected.",
    "maxInvalidLogin": {
      "time": 10,
      "count": 10,
      "coolofftime": 10
    },
    "watchDog": {
      "interval": 100,
      "timeout": 400
    }
  },
  "domaindefaults": {
    "__comment__": "Any settings in this section is used as default setting for all domains",
    "title": "meshcentral",
    "title2": "company",
    "_footer": "server.domain.name",
    "newAccounts": false
  },
  "domains": {
    "": {
      "siteStyle": 2,
      "titlePicture": "custom-logo.png",
      "loginPicture": "custom-logo.png",
      "minify": true,
      "newAccountEmailDomains": [
        "company.domain.name"
      ],
      "newAccountsRights": [
        "nonewgroups",
        "notools"
      ],
      "loginfooter": "This is a private server.",
      "certUrl": "https://server.domain.name:443/",
      "deviceMeshRouterLinks": {
        "rdp": true,
        "ssh": true,
        "scp": true,
        "_extralinks": [
          {
            "name": "HTTP",
            "protocol": "http",
            "port": 80,
            "_ip": "192.168.1.100",
            "_filter": [
              "mesh//xxxx",
              "node//xxxx"
            ]
          },
          {
            "name": "HTTPS",
            "protocol": "https",
            "port": 443
          }
        ]
      },
      "myServer": {
        "Backup": true,
        "Restore": false,
        "Upgrade": true,
        "ErrorLog": true,
        "Console": true,
        "Trace": false
      },
      "passwordRequirements": {
        "min": 8,
        "max": 128,
        "upper": 1,
        "lower": 1,
        "numeric": 1,
        "nonalpha": 1,
        "_reset": 90,
        "_force2factor": true,
        "skip2factor": "127.0.0.1",
        "oldPasswordBan": 3,
        "banCommonPasswords": true
      },
      "twoFactorCookieDurationDays": 30,
      "geoLocation": true,
      "novnc": true,
      "mstsc": true,
      "ssh": true,
      "consentMessages": {
        "title": "MeshCentral",
        "desktop": "{0} requesting remote desktop access. Grant access?",
        "terminal": "{0} requesting remote terminal access. Grant access?",
        "files": "{0} requesting remote files access. Grant access?"
      },
      "notificationMessages": {
        "title": "MeshCentral",
        "desktop": "{0} started a remote desktop session.",
        "terminal": "{0} started a remote terminal session.",
        "files": "{0} started a remote files session."
      },
      "agentCustomization": {
        "displayName": "Company AMT Agent",
        "description": "Company AMT agent for remote monitoring, management and assistance.",
        "companyName": "meshagent",
        "serviceName": "AMT Agent",
        "fileName": "meshagent"
      },
      "userConsentFlags": {
        "desktopnotify": false,
        "terminalnotify": false,
        "filenotify": false,
        "desktopprompt": false,
        "terminalprompt": false,
        "fileprompt": false,
        "desktopprivacybar": true
      },
      "urlSwitching": true,
      "desktopPrivacyBarText": "Your computer is being controlled by {0} ({1})",
      "amtManager": {
        "adminAccounts": [
          {
            "user": "admin",
            "pass": "<snip>"
          }
        ],
        "environmentDetection": [
          ""
        ],
        "wifiProfiles": [
          {
            "name": "<snip>",
            "ssid": "<snip>",
            "authentication": "wpa2-psk",
            "encryption": "ccmp-aes",
            "password": "<snip>"
          }
        ]
      },
      "_agentConfig": [
        "webSocketMaskOverride=1",
        "coreDumpEnabled=1"
      ],
      "authStrategies": {
        "azure": {
          "callbackurl": "https://server.domain.name/auth-azure-callback",
          "newAccounts": true,
          "clientid": "<snip>",
          "clientsecret": "<snip>",
          "tenantid": "<snip>"
        }
      }
    }
  }
}
```

The main web UI is running behind nginx, set up as per the install guide, but I'm not proxying / terminating TLS for AMS in nginx. It did occur to me to try that, but when I did it appeared to break CIRA, presumably because of the LetsEncrypt certificate nginx was presenting? (I may have not given it enough time).

Have I made some boneheaded mistake here, or is something broken? Happy to troubleshoot with whatever you'd like.

**Comment by nzalev** [comment]
I have submitted a fix for this. I don't personally have any agents to test this, but I believe it should be solved, so please try it out once it's live. Ylian is on vacation, so it may be a couple days before it's in a published version. Either version 0.8.66 or 0.8.67.

**Comment by Ylianst** [comment]
Wow, nice catch by @nzalev here! I just approved this fix, will be in MeshCentral v0.8.66.

**Comment by neg2led** [comment]
Awesome! Thanks for the fixup, updated my MeshCentral and it's working fine now 😄 

@nzalev no problem! most of my dayjob is dealing with vague problem tickets, so I try to avoid being That Guy - makes everyone's life a bit easier & usually gets things fixed faster, too 😝
